### ISSUE TITLE
Test execution exceptions are not supplied to extensions

### ISSUE THREAD

**Issue by sbrannen** [junit-team/junit5#92]
## Status Quo

`TestExtensionContext` does not provide access to any exception thrown during execution of the test method or other extensions.

## Deliverables

- [ ] Reintroduce a `getTestException()` method in the `TestExtensionContext` API.
  - Note: the prototype actually had such a [`getTestException()`](https://github.com/junit-team/junit-lambda/blob/prototype-1/junit5-api/src/main/java/org/junit/gen5/api/extension/TestExecutionContext.java#L37-L39) method.
- [ ] Ensure that the JUnit Jupiter `TestEngine` infrastructure tracks any exceptions encountered during test method or extension execution and supplies the exception to extensions via the `TestExtensionContext`.
  - Note: it is assumed that the engine already properly tracks the exceptions, but tests need to be introduced to verify the expected behavior.

## Further Resources

See the `afterEach()` method in the [SpringExtension](https://github.com/sbrannen/spring-test-junit5/blob/master/src/main/java/org/springframework/test/context/junit/jupiter/SpringExtension.java) for a concrete example of where the _test exception_ is required.

**Comment by jlink** [comment]
We must be very explicit about which exceptions will get there. For example: Should an extension context see exceptions occurred in other extension points and in other implementations of the same extension point? Or just from the same extension? Or just from the current test?

**Comment by sbrannen** [comment]
Good questions!

The `testException` should should in fact be an exception thrown during invocation of the test method. If we don't have clarity in this regard it would then be difficult if not impossible for an extension to properly react to such an exception. For example, if a developer wants to implement his own version of `@ExpectedException` using the JUnit 5 extension model, such an extension would need to access exactly the exception thrown from the test method and not from some other extension.

I tend to think that an exception thrown from a `@BeforeEach` or `@AfterEach` method should also be considered as a `testException`.

Regarding exceptions thrown by extensions, we could consider introducing an `extensionException` for that dedicated purpose. Clients of `TestExtensionContext` would then have to decide which of the two exceptions they need to support.

I think we should also consider introducing similar support for `ContainerExtensionContext`.

Thoughts?

**Comment by jlink** [comment]
Idea: Let's introduce a new extension point à la

interface ExcpetionThrownInTest extends ExtensionPoint {
   void exceptionThrown(TestExtensionContext context, Throwable throwable);
}

The callback would be invoked for all exceptions except AssertionError. Impementors could swallow or rethrow the exception.

Advantages:
- An extension can deal with exceptions in one place and not in all implemented extension points
- ExtensionContext need not be changed. 

Von meinem iPad gesendet

**Comment by sbrannen** [comment]
That unfortunately won't meet all of our needs.

Extensions that _wrap_ the test invocation with a _before_/_after_ callback pair need access to the exception in the _after_ callback.

For example, if a `DatabaseTransactionExtension` starts a transaction in the _before_ callback, it will either need to commit or roll back the transaction in the _after_ callback. In such cases, rollback behavior is typically influenced by the type of exception thrown.

In other words, _after_ callbacks will often need to process a test exception alongside the standard _after_ logic: separating _after_ logic and _exception handling_ logic into two different methods within an extension leads to cumbersome implementations at best.

**Comment by jlink** [comment]
Well, the exceptionThrown EP can store it as attribute for other EPs to pick up if need be.

**Comment by jlink** [comment]
This would play well together with my suggestion about `Store` in
https://github.com/junit-team/junit-lambda/issues/82

2016-01-09 15:34 GMT+01:00 Johannes Link <johannes.link@gmail.com>:

**Comment by sbrannen** [comment]
> Well, the exceptionThrown EP can store it as attribute for other EPs to pick up if need be.

Sure, it _could_, but... I don't see any need to make extension developers jump through hoops just to access the exception.

Accessing the exception should be just as easy as accessing the test class, instance, or method.

**Comment by sbrannen** [comment]
> This would play well together with my suggestion about `Store` in #82.

Let's look at that together next week!

**Comment by jlink** [comment]
Alright!

2016-01-09 19:08 GMT+01:00 Sam Brannen <notifications@github.com>:
